### PR TITLE
feat(health): fold successful healthchecks

### LINKED ISSUE #22796: checkhealth: fold "OK" (successful) sections by default
### Problem

With more neovim modules and plugins using checkhealth popping up, `:checkhealth` is starting to feel like reading CI logs with the amount of information that is presented to the user.

### Proposed solution

Automatically fold modules/sections that have passed their tests, but not for failed tests. This way the user will be shown the most important parts they need to know and fix, while also being able to inspect the information from the passing tests.

### PR DESCRIPTION
**Pull request by ateoi**
Fold successful health checks. This feature can be disable by setting the `g:checkhealth_folding_enable` to `v:false`.

Resolves #22796

### CODE REVIEW

**Comment by clason** [comment]
Thank you!

I'd prefer not to introduce a new config variable; just respect the global `fold` options -- either you like folds and want ALL THE THINGS folded, or you don't (and then nothing should be folded).

**Comment by ateoi** [comment]
> 
> 1. fold _all_ sections (failed or not)
> 2. indicate the status (fail, success, warning) via `foldtext`.
> 

Implemented the suggested behavior. If `foldenable`, all sections will be folded, and the checks with warning or error will be prefixed by `WE`.

checkhealth

**Comment by dundargoc** [comment]
>  either you like folds and want ALL THE THINGS folded, or you don't (and then nothing should be folded).

I disagree with this assessment. I don't want to block this PR, as I believe it's an improvement over the current situation, so I'm only saying this for future discussions. We can merge this as is.

The reason folds are controversial IMO is that what information should be hidden in a normal codebase is highly subjective. Simply put: we can't say with any certainty which parts of a codebase holds important information or not, so we can't be entirely sure whether we can safely hide code in folds or not. This is not the case with healthchecks: either the user has passed or failed the tests the plugin maintainer has deemed important.

If this is more of an objection to folds as how they work in practice, rather than a philosophical objection ("information should NEVER be hidden" kind of thing), then that is another matter entirely, and we can iterate on how to improve that (perhaps keymaps for healtcheck buffers, with keypresses determining the log level of the checks shown?).

**Comment by clason** [comment]
Hmm, I see your point, but don't fully agree with it. Let me break it down a bit:

1. Folds are controversial since "hidden text" is controversial (same as conceal; if I wanted invisible formatting characters, I'd be using Word). So _if_ there's folding (which I'm not against), it _must_ respect the global setting -- this is a red line for me.
2. Adding configuration options is not free, and I strongly believe they are not worth it here. Whether failed sections are folded or not is something I personally do not care about (since I'll never see it either way, see 1.) But as the point here is to maximize the amount of information at a glance, I do believe folding everything serves that purpose better. But maybe this can be tuned with different fold levels? There's also the question about how to treat warnings but not errors (or hints). 

Also, I think the case for folding becomes stronger once we have https://github.com/neovim/neovim/pull/20750, as then we can color-code the headings so you can see at a glance "all green" or "some red/yellow".

**Comment by dundargoc** [comment]
> So if there's folding (which I'm not against), it must respect the global setting -- this is a red line for me.

Yes, you're right. I was imagining folding as a mechanism for hiding information in this case rather than folding code, but that's a copout in hindsight. Folds are folds and must respect fold options.

> But as the point here is to maximize the amount of information at a glance, I do believe folding everything serves that purpose better.

Small nit: I'd say the point is to filter information rather than maximizing it, but I believe we essentially mean the same thing.

> But maybe this can be tuned with different fold levels?

I think this'd make sense. We already have a syntax files for the checkhealth buffer/filetype, perhaps these can be augmented with "sane" default folding behaviors (sane being my proposal naturally)?

> Also, I think the case for folding becomes stronger once we have https://github.com/neovim/neovim/pull/20750, as then we can color-code the headings so you can see at a glance "all green" or "some red/yellow".

Oooh, that looks nice.

In any case, I don't want to move the goalpost from this PR too much. I think this is perfectly fine for now and something we can iterate on in future PRs.

**Review by justinmk — APPROVED** [review]
Let's try this out, to unblock https://github.com/neovim/neovim/pull/23561

- resolve merge conflict
- add news.txt item
- add a small test case, or just update the existing test instead of setting `nofoldenable`

File: test/functional/plugin/health_spec.lua
Comment (justinmk) [inline test/functional/plugin/health_spec.lua:230]: should probably have a test case for folding

**Comment by ateoi** [comment]
> * add news.txt item

How can I update `news.txt` without breaking the tests?

**Comment by dundargoc** [comment]
You need to change `|foldenable|` to `|'foldenable'|`, since there is no tag without the ''.

File: runtime/lua/vim/health.lua
Comment (justinmk) [inline runtime/lua/vim/health.lua:21]:
```suggestion
    local err = '- ERROR '
    local errlen = string.len(err)
```

File: runtime/doc/news.txt
Comment (justinmk) [inline runtime/doc/news.txt:108]:
```suggestion
  defined by the 'foldenable' option.
```

**Comment by clason** [comment]
Hmm, would have preferred this to be opt-in (I forgot that `foldenable` is set by default.)

**Comment by clason** [comment]
No no, foldenable is going to be set to false, with a sledge hammer.

(I _hate_ folds with a passion.)

**Comment by justinmk** [comment]
`zR` expands all folds. For https://github.com/neovim/neovim/pull/23561 ,  "opt-out" is kind of required?

**Comment by clason** [comment]
So does `zi`; both add an extra step for using `checkhealth` as general diagnostics (which nvim-treesitter has started to lean into). `vim.o.foldenable = false` it is.

**Pull request merged**
